### PR TITLE
[cluster-test] decouple mempool and state sync msg sending and waiting

### PR DESCRIPTION
**Pull request by snakech**
decouple mempool and state sync sending msg and awaiting ack procedure to enhance the work load generation, compare with previous results, roughly 2x improvement.
from
```
2020-11-05T00:54:19.297740Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:209 Total mempool stats: exchanged 28685 messages, 1274015590 bytes
2020-11-05T00:54:19.297743Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:209 Average rate: exchanged 95 messages/s, 4246718 bytes/s
2020-11-05T00:54:19.297744Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:209 Total state sync stats: received 17767750 txs, exchanged 71073 messages, 2487555 bytes, 
2020-11-05T00:54:19.297746Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:209 Average rate: received 59225 txs/s, exchanged 236 msg/s, 8291 bytes/s,
```

to
```
2020-12-03T18:55:14.549851Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:152 Total mempool stats: exchanged 393614 messages, 5510596 bytes
2020-12-03T18:55:14.549865Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:153 Average rate: exchanged 1312 messages/s, 18368 bytes/s
2020-12-03T18:55:14.550702Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:161 Total state sync stats: received 33656990 txs, exchanged 35081 messages, 1263640 bytes, 
2020-12-03T18:55:14.550714Z [main] INFO testsuite/cluster-test/src/experiments/load_test.rs:162 Average rate: received 112189 txs/s, exchanged 116 msg/s, 4212 bytes/s, 
```
## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

(Share your test plan here. If you changed code, please provide us with clear instructions for verifying that your changes work.)

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/libra/developers.libra.org, and link to your PR here.)

### CODE REVIEW

File: testsuite/cluster-test/src/experiments/load_test.rs [hunk 1]
```diff
@@ -296,18 +299,23 @@ async fn mempool_load_test(
...
             transactions: vec![], // TODO submit actual txns
         };
-        // TODO log stats for bandwidth sent to remote peer to MempoolResult
-        bytes += lcs::to_bytes(&msg)?.len() as u64;
-        msg_num += 1;
-        sender.send_to(vfn, msg)?;
 
-        // await ACK from remote peer
-        let _response = events.select_next_some().await;
+        if pending < NETWORK_CHANNEL_SIZE {
```
Comment (zekun000) [inline testsuite/cluster-test/src/experiments/load_test.rs:309]: you probably should move the `let msg = ` inside this if

**Review by zekun000 — COMMENTED** [review]
also after the while loop ends, you should wait until all ack come back.

**Comment by snakech** [comment]
> also after the while loop ends, you should wait until all ack come back.

u are correct, will add another polling until thread is empty

**Review by zekun000 — COMMENTED** [review]
the two loops are similar, might worth abstracting it out.

File: testsuite/cluster-test/src/experiments/load_test.rs (hunk 1 above)
Comment (zekun000) [reply testsuite/cluster-test/src/experiments/load_test.rs:309]: ^^

**Comment by snakech** [comment]
> the two loops are similar, might worth abstracting it out.

I tried abstracting the whole loop but I think not really worth to do it at this point, I abstracted processing msg part for easier code reading

File: testsuite/cluster-test/src/experiments/load_test.rs (hunk 1 above)
Comment (zekun000) [reply testsuite/cluster-test/src/experiments/load_test.rs:310]: I don't think it's resolved?

File: testsuite/cluster-test/src/experiments/load_test.rs [hunk 2]
```diff
@@ -296,18 +300,23 @@ async fn mempool_load_test(
...
+            bytes += lcs::to_bytes(&msg)?.len() as u64;
+            sender.send_to(vfn, msg)?;
+            pending += 1;
+            msg_num += 1;
+        }
+        // await response from remote peer
+        if let Some(_response) = events.select_next_some().now_or_never() {
```
Comment (zekun000) [inline testsuite/cluster-test/src/experiments/load_test.rs:317]: this should be a while loop to poll all ready items?

File: testsuite/cluster-test/src/experiments/load_test.rs (hunk 1 above)
Comment (snakech) [reply testsuite/cluster-test/src/experiments/load_test.rs:310]: aha, sorry, I thought u meant let msg_num, resolved it now

File: testsuite/cluster-test/src/experiments/load_test.rs (hunk 2 above)
Comment (zekun000) [reply testsuite/cluster-test/src/experiments/load_test.rs:315]: this ^^

File: testsuite/cluster-test/src/experiments/load_test.rs [hunk 3]
```diff
@@ -296,18 +300,27 @@ async fn mempool_load_test(
...
+            pending += 1;
+        }
+        // await response from remote peer
+        while let Some(_response) = events.select_next_some().now_or_never() {
+            pending -= 1;
+            msg_num += 1;
+        }
+    }
+    // wait for all msgs are processed
+    time::delay_for(Duration::from_secs(10)).await;
```
Comment (zekun000) [inline testsuite/cluster-test/src/experiments/load_test.rs:321]: this is not the right way to do it
```
while pending > 0 {
  if let Some(_response) = events.next().await() {
    pending -= 1;
  }
}
```

File: testsuite/cluster-test/src/experiments/load_test.rs [hunk 4]
```diff
@@ -389,26 +418,35 @@ async fn state_sync_load_test(
...
+                served_txns += txs_num as u64;
+                pending -= 1;
+                msg_num += 1;
             }
         }
     }
+    // wait for all msgs are processed and acks back
+    loop {
+        let response = events.select_next_some().await;
+        if let Some(txs_num) = process_poll_msg(response) {
```
Comment (zekun000) [inline testsuite/cluster-test/src/experiments/load_test.rs:443]: no no no, this doesn't work, if it's some different messages coming back, you'll just break the loop. you need to use the pending counter

**Review by zekun000 — COMMENTED** [review]
can you try to run it and see what the number is? the NETWORK_CHANNEL_SIZE is 1024, and I saw MAX_CONCURRENT_NETWORK_REQS = 100, not sure if 1024 would work.

**Comment by snakech** [comment]
> can you try to run it and see what the number is? the NETWORK_CHANNEL_SIZE is 1024, and I saw MAX_CONCURRENT_NETWORK_REQS = 100, not sure if 1024 would work.

The numbers under PR description is the result with running this config, NETWORK_CHANNEL_SIZE = 1024, I can re-run it but I think we can change NETWORK_CHANNEL_SIZE down to 100 to match the MAX_CONCURRENT_NETWORK_REQS

**Comment by zekun000** [comment]
> The numbers under PR description is the result with running this config

do you run it after the last change (waiting for pending to be 0)? I'm not sure if some messages were dropped

**Comment by snakech** [comment]
> > The numbers under PR description is the result with running this config
> 
> do you run it after the last change (waiting for pending to be 0)? I'm not sure if some messages were dropped

I think this is the point we should to check, I am running the test but seems infra is flaky, will update here once I got numbers

**Pull request closed without merging**